src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,137 @@
+            add(value);
+            return;
+        }
+        Node<T> newNode = new Node<>(value);
+        if (index == 0) {
+            newNode.next = head;
+            head = newNode;
+            size++;
+            return;
+        }
+        Node<T> nextNode = getNodeByIndex(index);
+        Node<T> previousNode = nextNode.previous;
+        previousNode.next = newNode;
+        newNode.previous = previousNode;
+        newNode.next = nextNode;
+        nextNode.previous = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        list.forEach(this::add);
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = getNodeByIndex(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index == 0) {
+            T oldValue = head.value;
+            head = head.next;
+            size--;
+            return oldValue;
+        }
+        if (index >= 0 && index == size - LAST_ELEMENT_INDEX_SHIFT) {
+            T oldValue = tail.value;
+            tail = tail.previous;
+            size--;
+            return oldValue;
+        }
+        Node<T> nodeToRemove = getNodeByIndex(index);
+        nodeToRemove.previous.next = nodeToRemove.next;
+        nodeToRemove.next.previous = nodeToRemove.previous;
+        size--;
+        return nodeToRemove.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> node = head;
+        for (int i = 0; i < size; i++) {
+            if (object != null && node.value == null) {
+                node = node.next;
+                continue;
+            }
+            if (object == node.value || node.value.equals(object)) {
+                remove(i);
+                return true;
+            }
+            node = node.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        validateIndex(index);
+        Node<T> node = head;
+        for (int i = 0; i < index; i++) {
+            node = node.next;
+        }
+        return node;
+    }
+
+    private void validateIndex(int index) {
+        if (index >= size || index < 0) {
+            String format = String.format(INDEX_OUT_OF_BOUNDS_FORMAT, index, size);
+            throw new IndexOutOfBoundsException(format);
+        }
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> next;
+        private Node<T> previous;
+
+        public Node(T value) {
+            this.value = value;
+        }
```

Comment:
Private methods and classes should be after public ones in your class. Move the `Node` class to the bottom.